plugins/SegmentEditor/Model.php

```diff
@@ -129,6 +129,75 @@ public function getSegmentByDefinition($definition)
         return $segment;
     }
 
+    /**
+     * Gets a list of segments that have been deleted in the last week and therefore may have orphaned archives.
+     * @param Date $date Segments deleted on or after this date will be returned.
+     * @return array of segments. The segments are only populated with the fields needed for archive invalidation
+     * (e.g. definition, enable_only_idsite).
+     * @throws \Exception
+     */
+    public function getSegmentsDeletedSince(Date $date)
+    {
+        $dateStr = $date->getDatetime();
+        $sql = "SELECT DISTINCT definition, enable_only_idsite FROM " . Common::prefixTable('segment')
+            . " WHERE deleted = 1 AND ts_last_edit >= ?";
+        $deletedSegments = Db::fetchAll($sql, array($dateStr));
+
+        if (empty($deletedSegments)) {
+            return array();
+        }
+
+        $existingSegments = self::getExistingSegmnetsLike($deletedSegments);
+
+        foreach ($deletedSegments as $i => $deleted) {
+            $deletedSegments[$i]['idsites_to_preserve'] = array();
+            foreach ($existingSegments as $existing) {
+                if ($existing['definition'] != $deleted['definition']) {
+                    continue;
+                }
+
+                if (
+                    $existing['enable_only_idsite'] == $deleted['enable_only_idsite']
+                    || $existing['enable_only_idsite'] == 0
+                ) {
+                    // There is an identical segment (for either the specific site or for all sites) that is active
+                    // The archives for this segment will therefore still be needed
+                    unset($deletedSegments[$i]);
+                    break;
+                } elseif ($deleted['enable_only_idsite'] == 0) {
+                    // It is an all-sites segment that got deleted, but there is a single-site segment that is active
+                    // Need to make sure we don't erase the segment's archives for that particular site
+                    $deletedSegments[$i]['idsites_to_preserve'][] = $existing['enable_only_idsite'];
+                }
+            }
+        }
+
+        return $deletedSegments;
+    }
+
+    private static function getExistingSegmnetsLike(array $segments)
+    {
+        $whereClauses = array();
+        $bind = array();
+        foreach ($segments as $segment) {
+            $bind[] = $segment['definition'];
+            if ($segment['enable_only_idsite'] == 0) {
```

Comment:
@katebutler but concerned about this code here: https://github.com/matomo-org/matomo/blob/3.12.0-b3/core/Segment.php#L104-L108

Not sure if this can have any effect on the condition. I wonder if there are cases where we would store `actionType==sitesearches` as definition in the DB, and if passed differently in the API if we would also accept / store `"actionType%3D%3Dsitesearches` in case it is double encoded and not just once... according to the logic both segment definitions would be accepted in that class and have the same hash. The segment editor would probably store the segments always in the same way, but if someone uses the API and double encodes it, this might still work by the looks and we would store it urlencoded...

actually I just tested it and this would be the case:
![image](https://user-images.githubusercontent.com/273120/64655654-0a49f700-d481-11e9-8b4c-b48602f770d1.png)

See `https://example.com/index.php?module=API&method=SegmentEditor.add&format=JSON&token_auth=88d86212d1967a8565d05e51b1bfaa19&idSite=1&name=foo&definition=actionType%253D%253Dsitesearches`   vs using `actionType%3D%3Dsitesearches` as definition.

So it seems we also need to do `(definition = ? or definition = ? or definition = ?)` with a bind of `$definition, urlencode($definition), urldecode($definition)` for each segment :( because we can't know if the given definition is urlencoded or urldecoded etc.